unicorn/src/unicorn.cpp

```diff
@@ -0,0 +1,795 @@
+    try {
+        if(rc == ZNONODE) {
+            service->zk.create(path, value, ephemeral, sequence, *this);
+        }
+        else if (rc != ZOK) {
+            result.abort(rc, "Error during value get:" + zookeeper::get_error_message(rc));
+        }
+        else {
+            value_t parsed;
+            if (!value.empty()) {
+                parsed = unserialize(value);
+            }
+            if (stat.numChildren != 0) {
+                rc = zookeeper::ZOO_EXTRA_ERROR::CHILD_NOT_ALLOWED;
+                result.abort(rc, "Error during value get:" + zookeeper::get_error_message(rc));
+                return;
+            }
+            if (!parsed.is_double() && !parsed.is_int() && !parsed.is_null() && !parsed.is_uint()) {
+                rc = zookeeper::ZOO_EXTRA_ERROR::INVALID_TYPE;
+                result.abort(rc, "Error during value get:" + zookeeper::get_error_message(rc));
+                return;
+            }
+            if (parsed.is_double() || initial_value.is_double()) {
+                total = parsed.to<double>() + initial_value.to<double>();
+                service->zk.put(path, serialize(total), stat.version, *this);
+            }
+            else {
+                total = parsed.to<int64_t>() + initial_value.to<int64_t>();
+                service->zk.put(path, serialize(total), stat.version, *this);
+            }
+        }
+    }
+    catch(const zookeeper::exception& e) {
+        COCAINE_LOG_WARNING(service->log, "Failure during get action of increment: %s", e.what());
+        result.abort(e.code(), zookeeper::get_error_message(e.code()));
+    }
+}
+
+void
+unicorn_dispatch_t::increment_action_t::finalize(zookeeper::value_t) {
+    result.write(versioned_value_t(initial_value, version_t()));
+}
+
+void
+unicorn_dispatch_t::increment_action_t::abort(int rc) {
+    result.abort(rc, zookeeper::get_error_message(rc));
+}
+
+distributed_lock_t::lock_action_t::lock_action_t(
+    const zookeeper::handler_tag& tag,
+    unicorn_service_t* service,
+    std::shared_ptr<distributed_lock_t::lock_state_t> _state,
+    path_t _path,
+    path_t _folder,
+    value_t _value,
+    unicorn_dispatch_t::response::lock _result
+) :
+    zookeeper::managed_handler_base_t(tag),
+    unicorn_dispatch_t::create_action_base_t(tag, service, std::move(_path), std::move(_value), true, true),
+    zookeeper::managed_strings_stat_handler_base_t(tag),
+    zookeeper::managed_stat_handler_base_t(tag),
+    zookeeper::managed_watch_handler_base_t(tag),
+    state(std::move(_state)),
+    result(std::move(_result)),
+    folder(std::move(_folder))
+{}
+
+void
+distributed_lock_t::lock_action_t::operator()(int rc, std::vector<std::string> childs, zookeeper::node_stat const& stat) {
+    /**
+    * Here we assume:
+    *  * nobody has written trash data(any data except locks) in specified directory
+    *  * sequence number issued by zookeeper is comaparable by string comparision.
```

Comment:
It's client responsibility not to use path specified as lock path to store any data in it's children. However most dumb cases(non sequence type nodes) we can handle